authorization-common/src/test/java/com/sequenceiq/authorization/EnforceAuthorizationLogicsUtil.java

```diff
@@ -62,18 +124,239 @@ public static void testIfControllerClassHasAuthorizationAnnotation() {
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .map(Object::toString)
+                    .collect(toList());
+        }
+    }
+
+    private static String invalid(Method method, String reason) {
+        return methodToString(method) + ": " + reason;
+    }
+
+    private static Function<Method, Optional<String>> noRestriction() {
+        return method -> Optional.empty();
+    }
+
+    private static Function<Method, Optional<String>> unknownAnnotation(Class<? extends Annotation> annotation) {
+        return method -> Optional.of(invalid(method, "No validation rule specified for " + annotation));
+    }
+
+    private static String methodToString(Method method) {
+        return method.getDeclaringClass().getSimpleName() + '#' + method.getName();
+    }
+
+    private static Function<Method, Optional<String>> hasParamWhere(Class<? extends Annotation> annotation,
+            Function<Pair<Method, Class<?>>, Optional<String>> typeValidator) {
+        return method -> {
+            List<Optional<String>> validations = new ArrayList<>();
+            for (Parameter parameter : method.getParameters()) {
+                if (parameter.isAnnotationPresent(annotation)) {
+                    validations.add(typeValidator.apply(Pair.of(method, parameter.getType())));
+                }
+            }
+            String errorMessageCommon = " method param with @" + annotation.getSimpleName() + " annotation";
+            if (validations.isEmpty()) {
+                return Optional.of(invalid(method, "Misssing" + errorMessageCommon));
+            } else if (validations.size() > 1) {
+                return Optional.of(invalid(method, "Multiple" + errorMessageCommon));
+            } else {
+                return validations.get(0);
+            }
+        };
+    }
+
+    private static Function<Method, Optional<String>> stringParam(Class<? extends Annotation> annotation) {
+        return hasParam(annotation, String.class);
+    }
+
+    private static Function<Method, Optional<String>> hasParam(Class<? extends Annotation> annotation, Class<?> type) {
+        return method -> {
+            int count = 0;
+            for (Parameter parameter : method.getParameters()) {
+                if (parameter.isAnnotationPresent(annotation) && type.equals(parameter.getType())) {
+                    count++;
+                }
+            }
+            String errorMessageCommon = " method parameter with @" + annotation.getSimpleName() + " annotation and type " + type.getSimpleName();
+            if (count == 0) {
+                return Optional.of(invalid(method, "Misssing" + errorMessageCommon));
+            } else if (count > 1) {
+                return Optional.of(invalid(method, "Multiple" + errorMessageCommon));
+            } else {
+                return Optional.empty();
+            }
```

Comment:
as I see this code block appears twice in the class, please extract it

Reply:
ok